### ISSUE TITLE
port twisted.names.srvconnect to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8262]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)| @pawelmhm reported|
|-|-|
|Trac ID|trac#8262|
|Type|enhancement|
|Created|2016-03-29 19:09:10Z|
|Branch|https://github.com/twisted/twisted/tree/py3-srvconnect-8262|

See previous tickets for porting names:

[#6093](https://github.com/twisted/twisted/issues/6093) [#6056](https://github.com/twisted/twisted/issues/6056) [#6059](https://github.com/twisted/twisted/issues/6059) [#6057](https://github.com/twisted/twisted/issues/6057) [#6058](https://github.com/twisted/twisted/issues/6058) [#6092](https://github.com/twisted/twisted/issues/6092) [#8195](https://github.com/twisted/twisted/issues/8195), [#8259](https://github.com/twisted/twisted/issues/8259)

There are some old tickets for srvconnect: 

one about cleanup [#2800](https://github.com/twisted/twisted/issues/2800)
other about removing cmp usage [#5124](https://github.com/twisted/twisted/issues/5124)

<details><summary>Searchable metadata</summary>

```
trac-id__8262 8262
type__enhancement enhancement
reporter__pawelmhm pawelmhm
priority__normal normal
milestone__Python_3_x Python-3.x
branch__py3_srvconnect_8262 py3-srvconnect-8262
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__None None
time__1459278550427084 1459278550427084
changetime__1465913311552501 1465913311552501
version__None None
owner__ralphm ralphm

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to @ralphm<br>@ralphm set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm removed owner<br>@ralphm set status to `new`|
|-|-|

Builders running. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @ralphm|
|-|-|

Changes looks good. Many thanks!

0. There is a twistedchecker error :)

1. Do we have coverage for new code from `Record_SRV.__lt__` ? I can not see it by reading the code, and codecov is also reporing a missing coverage https://codecov.io/gh/twisted/twisted/commit/6065a98bae885115647ce6dc3231b57ccd79578d

2. Why use nativeString for twisted.names.client.lookupService ?

I saw your commit in which you added nativeString, but I think that is the wrong way to fix it.

Maybe this can be fixed in a separate ticket, but we should add a comment referencing that ticket.

I was expecting to either go for an text friend Unicdoe based api and use Unicode on both py2 and py3, or go for low level bytes in both py2 and py3.

---------

codecov.is broken right now, but we should make sure that we have 100% coverage for the ported code.

Please address the comments and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm removed owner|
|-|-|

Replying to [adiroiban](#note_3):
> Changes looks good. Many thanks!
> 
> 0. There is a twistedchecker error :)

I don't understand this at all. What constitutes an error here? I always see:

```ERROR:   twistedchecker: commands failed```

> 1. Do we have coverage for new code from `Record_SRV.__lt__` ? I can not see it by reading the code, and codecov is also reporing a missing coverage https://codecov.io/gh/twisted/twisted/commit/6065a98bae885115647ce6dc3231b57ccd79578d

Added tests. Thanks for this, because the comparison was wrong.

> 2. Why use nativeString for twisted.names.client.lookupService ?
> 
> I saw your commit in which you added nativeString, but I think that is the wrong way to fix it.
> 
> Maybe this can be fixed in a separate ticket, but we should add a comment referencing that ticket.
> 
> I was expecting to either go for an text friend Unicdoe based api and use Unicode on both py2 and py3, or go for low level bytes in both py2 and py3.

I had to change this because bytes formatting was only introduced in Python 3.5. Because `.encode('idna')` is called on `SRVConnector.domain` (in its `__init__`) as well as on `name` in `dns.Name` (which is what the argument to lookupService is wrapped in), and both `service` and `protocol` are ASCII-only, all components are ASCII-only. Using `nativeString` was the easiest to make the formatting work properly, while ensuring that all things are indeed ASCII.

By the way, I forgot to actually create a pull request, so that's here: https://github.com/twisted/twisted/pull/91.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set owner to @ralphm|
|-|-|

Hi Ralph, thanks for working on this.

1. I believe Adi is talking about https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/106/steps/run-twistedchecker/logs/new%20twistedchecker%20errors for the twistedchecker error.
1. https://github.com/twisted/twisted/pull/91/files#diff-6ffb5ed60a86133187f8712b2c14f7eaR36 should be a decorator. Could you also write a test to verify that SRVConnector implements this interface?
1. https://github.com/twisted/twisted/pull/91/files#diff-6ffb5ed60a86133187f8712b2c14f7eaR156 needs spaces on each side of the =.
1. New docstrings in the tests need ```L{}/C{```} around Python identifiers.

Please fix these and submit for rereview. Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm removed owner|
|-|-|

Replying to [hawkowl](#note_5):
> Hi Ralph, thanks for working on this.
> 
> 1. I believe Adi is talking about https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/106/steps/run-twistedchecker/logs/new%20twistedchecker%20errors for the twistedchecker error.

Ah. This was not obvious to me. Thanks!

> 2. https://github.com/twisted/twisted/pull/91/files#diff-6ffb5ed60a86133187f8712b2c14f7eaR36 should be a decorator. Could you also write a test to verify that SRVConnector implements this interface?
> 3. https://github.com/twisted/twisted/pull/91/files#diff-6ffb5ed60a86133187f8712b2c14f7eaR156 needs spaces on each side of the =.
> 4. New docstrings in the tests need ```L{}/C{```} around Python identifiers.

Fixed.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to ralph|
|-|-|

Thanks ralphm!  Unfortunately I do have some additional feedback here:

1. I'm a little concerned about adding ordering, which is new and subtle public API, to `Record_SRV`, when the real thing you need is simply to switch from `cmp=` to `key=` in your `sort=` parameter.  Consider the much simpler strategy of just using [https://docs.python.org/3.5/library/functools.html#functools.cmp_to_key `cmp_to_key`] instead.
2. If you are going to add ordering, I realize (after consulting the reference documentation) that `sort()` only uses `<` and so that's technically all you need, but this will be confusing to users of the class, especially since presumably the behavior of other kinds of comparison operation would be added later.  The `@functools.total_ordering` decorator can make a class fully ordered with what's already provided here, but it needs a lot more tests; plus, `__lt__` would need a docstring and a `@since` field.

I'd probably prefer it if we just stuck to changing the parameter to `sort`, especially given that the `cmp_to_key` shortcut exists on all relevant pythons (including 2.7!).

Please address and resubmit.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Oh, oops. I meant to say in point 2 there that if you do intend to add ordering, you should probably add ordering in a separate PR first given that it is going to need a few dozen more tests to check the various comparability operators on `Record_SRV` first.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm removed owner|
|-|-|

I reverted the addition of `__lt__` and now just use a lambda that returns `(priority, weight)` as the key for sort.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set owner to ralph|
|-|-|

Here is my review

1. `Two records with equal priorities compare on weight.` compare, how? for non-DNS experts, just by reading the docstring is hard to tell of the comparison. Should the small value come first?

Also, by reading the test I don't see how this things work... you set some list in orderedServers and I can see that repeated calls to pickServer() will return the list in the same order :)

pickServer has no docstring :(

In a similar way, the docstring from test_srvDifferentPriorities could be improved.

2. test_SRVLookupName talks about bytes, but the tests uses nativeString

3. Why do need to update twisted/test/proto_helpers.py ?

Please check my comments and resubmit.

I think that the docstring of the test should be expanded to describe the test scenario and the expected result.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_11"></a>|@ralphm removed owner|
|-|-|

Replying to [adiroiban](#note_10):
> Here is my review
> 
> 1. `Two records with equal priorities compare on weight.` compare, how? for non-DNS experts, just by reading the docstring is hard to tell of the comparison. Should the small value come first?
> 
> Also, by reading the test I don't see how this things work... you set some list in orderedServers and I can see that repeated calls to pickServer() will return the list in the same order :)
> 
> pickServer has no docstring :(
> 
> In a similar way, the docstring from test_srvDifferentPriorities could be improved.

There are a/were a couple of things particularly wrong with SRVConnector:

 1. It didn't use any random number, as required by RFC 2782
 2. When a connection attempt fails, it will not continue to the next one, but stops the connector, which in turn stops the protocol factory.
 3. `ReconnectingClientFactory`, which is currently used by `XmlStreamFactory` implementation in the Jabber/XMPP code, ignores the call to stop the factory and retries anyway, so I think this magically works initially. However, in this case, if you have a long-lived connection which drops, `SRVConnector` should start over, not continue with the next server.

I fixed 1 now as part of this branch, so that we at least have a decent implementation of the algorithm.

Regarding 2 and 3, the use of `ReconnectingClientFactory` should be abandoned. Instead all of this should use `twisted.application.internet.ClientService`. I had a short discussion with Glyph on this at Pycon. It might be we have to do something with the same interface, but tailored to how SRV resolution should work. In any case this all requires a lot of changes all over the XMPP implementation, not only here but also in Wokkel.

> 2. test_SRVLookupName talks about bytes, but the tests uses nativeString

Woops, this was a left-over from an earlier version.
 
> 3. Why do need to update twisted/test/proto_helpers.py ?

You're right. I've now find that this change really belongs in `SRVConnector` itself, so I changed it to that effect and reversed this part of the change.
 
> Please check my comments and resubmit.
> 
> I think that the docstring of the test should be expanded to describe the test scenario and the expected result.
> 
> Thanks!

Thanks for your comments. This should be a lot better. Bots have run.

Please review.

B.t.w. it turns out that this branch also fixes [#5124](https://github.com/twisted/twisted/issues/5124) and [#5348](https://github.com/twisted/twisted/issues/5348).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl set owner to @ralphm|
|-|-|

Hi Ralph, thanks for this.

1. Could you add return type documentation to ```pickServer```?
1. The str() in ```pickServer``` concerns me -- I think that it might do the wrong thing, depending on what the type is -- is it safe to omit it? Do we want to use nativeString instead?

Please look at these two points, then merge away. (The git migration should mean it's just a simple ```git merge``` now :) ) Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_13"></a>|@ralphm commented|
|-|-|

Replying to [hawkowl](#note_12):
> Hi Ralph, thanks for this.
> 
> 1. Could you add return type documentation to ```pickServer```?

Done.

> 1. The str() in ```pickServer``` concerns me -- I think that it might do the wrong thing, depending on what the type is -- is it safe to omit it? Do we want to use nativeString instead?

So `chose.target` is a `twisted.names.dns.Name` here, which has to be converted to a string, and it defines `__str__` for this because of the following. If it makes you feel better, I can always call `nativeString` on it explicitly, but it doesn't matter much here.

The first element of the tuple returned by `pickServer` is passed to `connectFunc` in `_reallyConnect`, which is `connectTCP`. Although the interface for `connectTCP` documents its `host` argument as `bytes`, the implementation (eventually through `twisted.internet.tcp._BaseTCPClient.__init__`) will call `isIPAddress` and `isIPv6Address` from `twisted.internet.abstract`. Each of these expect native strings, as they do string splitting or use the `in` operator with native string literals. Their docstrings say `C{str}` which might indicate nobody payed close attention to them lately, but I am not sure which of these is wrong. Either `connectTCP`'s interface should explicitly allow native strings, or the IP test functions should only accept bytes.

Opinions?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_14"></a>|@ralphm commented|
|-|-|

Based on IRC discussion with glyph and hawkowl, we conclude that `connectTCP` should really accept (at least) ASCII-only native strings. I filed a ticket for this here: [#8470](https://github.com/twisted/twisted/issues/8470).

I believe that concludes the review comments from hawkowl, so I'll go ahead and merge this.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_15"></a>|@ralphm set status to `closed`|
|-|-|

In changeset 5aee182293607ad4b806d81c26dbbd1f6e1f4624
```
#!CommitTicketReference repository="" revision="5aee182293607ad4b806d81c26dbbd1f6e1f4624"
Merge py3-srvconnect-8262: Port twisted.names.srvconnect to Python 3.

Author: ralphm
Reviewer: adiroiban, hawkowl, glyph
Fixes: #8262, #5124, #2800

This also cleans up the use of a cmp function for sorting SRV records,
makes the `pickServer` implementation in line with the algorithm
specified in the RFC and adds tests around that to match.
```
